### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by mereacre**
Added edgesec integration test.

Please check with the git rebase. It seems that there are problems.

### CODE REVIEW

**Comment by aloisklink** [comment]
I'm not even sure what's happened with your git history again!

Did you do some sort of command to sign your commits? My gut feeling is that you did a `git rebase --gpg-sign` or something? But somehow, you even have some commits in triplicate, so I really don't know how you managed this :laughing: 

GitHub is only showing that you've changed a few files, which seems wrong, so I'll try to review only once the commits are fixed up.

To fix, can you try doing a `git rebase --interactive --sign-off origin/main`, then `drop` every commit that's already in `main`.

E.g., something like:

image

Git docs here: https://git-scm.com/docs/git-rebase#_interactive_mode in case it helps.

**Comment by aloisklink** [comment]
Commits have been rebased. @mereacre, can you review my changes? I slightly modified your code. I'll do a full review tomorrow.

**Review by mereacre — COMMENTED** [review]
Looks good.

File: Testing/Temporary/CTestCostData.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+---
```
[the remark this answers is not in the document]
Comment (mereacre) [reply Testing/Temporary/CTestCostData.txt]: Done

File: Testing/Temporary/LastTest.log [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Start testing: Nov 04 17:34 GMT
+----------------------------------------------------------
+End testing: Nov 04 17:34 GMT
```
[the remark this answers is not in the document]
Comment (mereacre) [reply Testing/Temporary/LastTest.log]: Done

File: src/ap/hostapd.c [hunk 3]
```diff
@@ -46,18 +46,18 @@ static char hostapd_proc_name[MAX_OS_PATH_LEN];
...
-    log_errno("unlink");
+  FILE *fp = fopen(vlan_file, "w");
+
+  if (fp == NULL) {
+    log_errno("fopen");
     return -1;
   }
 
-  FILE *fp = fopen(vlan_file, "a+");
+  fp = freopen(vlan_file, "a+", fp);
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/ap/hostapd.c]: Done

File: src/ap/hostapd.c [hunk 4]
```diff
@@ -116,18 +116,16 @@ int generate_hostapd_conf(struct apconf *hconf, struct radius_conf *rconf) {
 }
 #else
 int generate_hostapd_conf(struct apconf *hconf, struct radius_conf *rconf) {
-  // Delete the config file if present
-  int stat = unlink(hconf->ap_file_path);
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/ap/hostapd.c:120]: Done

File: src/ap/hostapd.c [hunk 5]
```diff
@@ -116,18 +116,16 @@ int generate_hostapd_conf(struct apconf *hconf, struct radius_conf *rconf) {
...
-  if (stat == -1 && errno != ENOENT) {
-    log_errno("unlink");
+  FILE *fp = fopen(hconf->ap_file_path, "w");
+  if (fp == NULL) {
+    log_errno("fopen");
     return -1;
   }
 
-  FILE *fp = fopen(hconf->ap_file_path, "a+");
+  fp = freopen(hconf->ap_file_path, "a+", fp);
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/ap/hostapd.c]: Done.

File: src/config.c [hunk 6]
```diff
@@ -730,81 +730,81 @@ bool load_firewall_config(const char *filename, struct firewall_conf *config) {
...
     log_errno("Couldn't open %s config file.\n", filename);
-    return false;
+    return -1;
   }
   fclose(fp);
 
   os_strlcpy(config->config_ini_path, filename, MAX_OS_PATH_LEN);
 
   if (!load_system_config(filename, config)) {
     log_debug("load_system_config fail");
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/config.c:745]: fopen sets the errno. The purpose of errno is to catch standard library errors.

File: src/dhcp/dnsmasq.c [hunk 7]
```diff
@@ -239,18 +238,17 @@ int generate_dnsmasq_conf(struct dhcp_conf *dconf, UT_array *dns_server_array) {
...
 
-  if (stat == -1 && errno != ENOENT) {
-    log_errno("unlink");
+  if (fp == NULL) {
+    log_errno("fopen");
     return -1;
   }
 
-  FILE *fp = fopen(dhcp_script_path, "a+");
+  fp = freopen(dhcp_script_path, "a+", fp);
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/dhcp/dnsmasq.c]: Done

File: src/firewall/firewall_service.c [hunk 8]
```diff
@@ -125,32 +127,34 @@ struct fwctx *fw_init_context(hmap_if_conn *if_mapper,
...
-    log_error("uwrt_cleanup_firewall fail");
-    fw_free_context(fw_ctx);
-    return NULL;
-  }
-
-  while ((p = (config_ifinfo_t *)utarray_next(config_ifinfo_array, p)) !=
-         NULL) {
-    if (uwrt_gen_firewall_zone(fw_ctx->ctx, p->brname) < 0) {
+  if (exec_firewall) {
+    config_ifinfo_t *p = NULL;
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/firewall/firewall_service.c]: Done

File: tests/data/uci/dhcp [hunk 9]
```diff
@@ -36,3 +36,58 @@ config odhcpd 'odhcpd'
 	option leasefile '/tmp/hosts/odhcpd'
 	option leasetrigger '/usr/sbin/odhcpd-update'
 	option loglevel '4'
+
+config dhcp '0'
```
[the remark this answers is not in the document]
Comment (mereacre) [reply tests/data/uci/dhcp:40]: Nope. It does change when I run the tests. Looks like we have to look at the tests again.

File: tests/data/uci/wireless [hunk 10]
```diff
@@ -46,3 +46,32 @@ config wifi-iface 'guest_iface_0'
 config wifi-iface 'guest_iface_1'
 	option disabled '1'
 	option device 'radio1'
+
+config wifi-iface 'edgesec'
```
[the remark this answers is not in the document]
Comment (mereacre) [reply tests/data/uci/wireless:50]: I reopened the branch that fixes the problem.

File: tests/test_edgesec.c [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+  struct radius_client_data *radius =
+      radius_client_init(eloop, /*ctx*/ eloop, &servers);
+  assert_non_null(radius);
+
+  ret =
+      radius_client_register(radius, RADIUS_AUTH, receive_auth, /*ctx*/ eloop);
+  assert_int_equal(ret, 0);
+
+  char buf[20];
+  static uint8_t addr[6] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05};
```
[the remark this answers is not in the document]
Comment (mereacre) [reply tests/test_edgesec.c]: Done

File: src/config.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mereacre) [reply src/config.c:745]: Good point about this one. It should be log_error.

**Comment by aloisklink** [comment]
Hi @mereacre, before you merge, can you run:

```bash
git rebase -i --autosquash 7cc308177d57e237c0dae7a5cf4808af641c5bd5 --rebase-merges
```

(no need to edit anything too)

It should automatically combine the [feat: Added execute firewall for uci](https://github.com/nqminds/edgesec/pull/339/commits/6d16f556f5ff338f3cd1928da5458d4829ac417d) commit with [fixup! feat: Added execute firewall for uci](https://github.com/nqminds/edgesec/pull/339/commits/5809992b6d87bd61ed0842eabf001ad7d5e12e14) (plus, it will allow you to sign all your commits since then and remove my face as the committer)

**Comment by aloisklink** [comment]
Hi @mereacre, I might need to wait until Monday to `git rebase` this PR again.

I'm swear there must be something different with your `git config` that is unusual!

Doing a `git reset --hard eb708cce43f9f5bd17d69bd2dd5ec36facbde760` might help, since https://github.com/nqminds/edgesec/pull/339/commits/eb708cce43f9f5bd17d69bd2dd5ec36facbde760 looks like the last good commit (most of everything after that looks like mistaken/broken commits).

**Comment by aloisklink** [comment]
@mereacre, it looks like the `test_edgesec` is failing due to `ERROR runctl.c:440: load_crypt_service fail` on the `linux-with-crypt` preset.

**Pull request merged**